File: src/leetcode/2025/12/Solution383.java
public class Solution383 {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] charCount = new int[26];

        for (char c : magazine.toCharArray()) {
            charCount[c - 'a']++;
        }

        for (char c : ransomNote.toCharArray()) {
            charCount[c - 'a']--;
            if (charCount[c - 'a'] < 0) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        Solution383 solution = new Solution383();

        String ransomNote1 = "a";
        String magazine1 = "b";
        boolean result1 = solution.canConstruct(ransomNote1, magazine1);
        System.out.println("Can construct ransom note: " + result1); // Expected output: false

        String ransomNote2 = "aa";
        String magazine2 = "ab";
        boolean result2 = solution.canConstruct(ransomNote2, magazine2);
        System.out.println("Can construct ransom note: " + result2); // Expected output: false

        String ransomNote3 = "aa";
        String magazine3 = "aab";
        boolean result3 = solution.canConstruct(ransomNote3, magazine3);
        System.out.println("Can construct ransom note: " + result3); // Expected output: true
    }
}

File: src/leetcode/2025/12/Solution383.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

class Solution383 {
public:
    bool canConstruct(string ransomNote, string magazine) {
        vector<int> charCount(26, 0);

        for (char c : magazine) {
            charCount[c - 'a']++;
        }

        for (char c : ransomNote) {
            charCount[c - 'a']--;
            if (charCount[c - 'a'] < 0) {
                return false;
            }
        }

        return true;
    }
};

int main() {
    Solution383 solution;
    
    string ransomNote1 = "a";
    string magazine1 = "b";
    bool result1 = solution.canConstruct(ransomNote1, magazine1); // Expected: false
    cout << (result1 ? "true" : "false") << endl;

    string ransomNote2 = "aa";
    string magazine2 = "ab";
    bool result2 = solution.canConstruct(ransomNote2, magazine2); // Expected: false
    cout << (result2 ? "true" : "false") << endl;
    
    string ransomNote3 = "aa";
    string magazine3 = "aab";
    bool result3 = solution.canConstruct(ransomNote3, magazine3); // Expected: true
    cout << (result3 ? "true" : "false") << endl;

    return 0;
}